app/swift_demangle.c

```diff
@@ -0,0 +1,147 @@
+
+#include "swift_demangler.h"
+#include <stdlib.h>
+#include <string.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <limits.h>
+#include <dlfcn.h>
+#include <unistd.h>
+
+// Function pointer type for swift_demangle.
+typedef char *(*swift_demangle_func_t)(const char *mangledName, size_t mangledNameLength,
+                                         char *outputBuffer, size_t *outputBufferSize,
+                                         uint32_t flags);
+
+// Static variables to hold the library handle and function pointer.
+static swift_demangle_func_t swift_demangle_ptr = NULL;
+static void *swift_demangle_handle = NULL;
+
+static int init_swift_demangle(void) {
+    if (swift_demangle_ptr)
+        return 1; // Already initialized.
+
+    // Get the PATH environment variable.
+    char *path_env = getenv("PATH");
+    if (!path_env) {
+        return 0;
+    }
+
+    // Duplicate PATH because strtok_r modifies the string.
+    char *path_dup = strdup(path_env);
+    if (!path_dup) {
+        return 0;
+    }
+
+    // Search for "swiftc" in each directory of PATH.
+    char swift_path[PATH_MAX];
+    int found = 0;
+    char *saveptr = NULL;
+    for (char *dir = strtok_r(path_dup, ":", &saveptr);
+         dir != NULL;
+         dir = strtok_r(NULL, ":", &saveptr)) {
+        snprintf(swift_path, sizeof(swift_path), "%s/swift", dir);
+        if (access(swift_path, X_OK) == 0) {
+            found = 1;
+            break;
+        }
+    }
+    free(path_dup);
+    if (!found) {
+        return 0;
+    }
+
+    // Expect swift_path to be of the form SOMEPATH/usr/bin/swift.
+    // Locate the "/usr/bin/swift" substring.
+    const char *needle = "/usr/bin/swift";
+    char *substr = strstr(swift_path, needle);
+    if (!substr) {
+        return 0;
+    }
+
+    // The installation root is everything preceding "/usr/bin/swift".
+    size_t base_len = substr - swift_path;
+    char base_path[PATH_MAX];
+    if (base_len >= sizeof(base_path)) {
+        return 0;
+    }
+    strncpy(base_path, swift_path, base_len);
+    base_path[base_len] = '\0';
+
+    // Construct the full path to libswiftDemangle.so:
+    //   SOMEPATH/usr/lib/libswiftDemangle.so
+    char lib_path[PATH_MAX];
+    snprintf(lib_path, sizeof(lib_path), "%s/usr/lib/libswiftDemangle.so", base_path);
```

Comment:
this seems like it's replicating parts of what we already do for the other demanglers, why can we not integrate with that code? why is the fancy logic to find swift needed, can't we just use normal library lookup instead?

Reply:
I thought that the upside of doing it this way was to make it future proof as swift seems still to be evolving fairly rapidly. The downside is that it requires swift to be in the PATH, but only for people that are interested in swift demangling. Other developers are not affected. Also I wasn't sure about licensing issues with distributing a binary library libswiftDemangle.so which is Apache licensed in a GPL project.

Reply:
IANAL but linking gpl to apache is fine but not the other way around?

still, you haven't answered my question - PATH has no impact on the dynamic library loader on linux - or are you saying this proposed approach here needs it? I would instead rather we try to `dlopen` `libswiftDemangle.so` directly, just like we do for `rustc_demangle` e.g. unless this isn't possible (why?) please simplify your patch accordingly.